File: week13/15/5_Fri_MaxDiffBT.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:        
        def dfs(node, small, big):
            if not node:
                return big - small
            small = min(node.val, small)
            big = max(node.val, big)
            left = dfs(node.left, small, big)
            right = dfs(node.right, small, big)
            return max(left, right)
        return dfs(root, root.val, root.val)

File: week13/15/5_Fri_MaxDiffBT.cpp
#include <algorithm>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        return dfs(root, root->val, root->val);
    }
    int dfs(TreeNode* node, int small, int big) {
        if (!node) return big - small;
        small = min(node->val, small);
        big = max(node->val, big);
        int left = dfs(node->left, small, big);
        int right = dfs(node->right, small, big);
        return max(left, right);
    }
};